src/main/java/com/yelp/nrtsearch/server/luceneserver/search/query/multifunction/GeoPointDecayFilterFunction.java

```diff
@@ -0,0 +1,158 @@
+
+import com.google.type.LatLng;
+import com.yelp.nrtsearch.server.grpc.MultiFunctionScoreQuery;
+import com.yelp.nrtsearch.server.luceneserver.doc.LoadedDocValues;
+import com.yelp.nrtsearch.server.luceneserver.geo.GeoPoint;
+import com.yelp.nrtsearch.server.luceneserver.geo.GeoUtils;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import org.apache.lucene.index.*;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.Query;
+
+public class GeoPointDecayFilterFunction extends DecayFilterFunction {
+
+  private final MultiFunctionScoreQuery.DecayFunction decayFunction;
+  private final String fieldName;
+  private final DecayFunction decayFunc;
+  private final double scale;
+  private final double offset;
+  private final double decay;
+  private final LatLng origin;
+
+  /**
+   * Constructor.
+   *
+   * @param filterQuery filter to use when applying this function, or null if none
+   * @param weight weight multiple to scale the function score
+   * @param decayFunction to score a document with a function that decays depending on the distance
+   *     between an origin point and a numeric doc field value
+   */
+  public GeoPointDecayFilterFunction(
+      Query filterQuery, float weight, MultiFunctionScoreQuery.DecayFunction decayFunction) {
+    super(filterQuery, weight, decayFunction);
+    this.decayFunction = decayFunction;
+    this.fieldName = decayFunction.getFieldName();
+    this.decayFunc = getDecayFunction(decayFunction.getDecayType());
+    this.offset = GeoUtils.getDistance(decayFunction.getOffset());
+    this.decay = decayFunction.getDecay();
+    double userGivenScale = GeoUtils.getDistance(decayFunction.getScale());
+    this.scale = decayFunc.computeScale(userGivenScale, decay);
+    if (!decayFunction.hasGeoPoint()) {
+      throw new IllegalArgumentException("Decay Function should have a geoPoint for Origin field");
+    } else {
+      this.origin = decayFunction.getGeoPoint();
+    }
+  }
+
+  @Override
+  public LeafFunction getLeafFunction(LeafReaderContext leafContext) throws IOException {
+    return new GeoPointDecayLeafFunction(leafContext);
+  }
+
+  public final class GeoPointDecayLeafFunction implements LeafFunction {
+
+    private final LoadedDocValues.SingleLocation geoPointValue;
+
+    public GeoPointDecayLeafFunction(LeafReaderContext context) throws IOException {
+      this.geoPointValue =
+          new LoadedDocValues.SingleLocation(context.reader().getSortedNumericDocValues(fieldName));
+    }
+
+    @Override
+    public double score(int docId, float innerQueryScore) throws IOException {
+      GeoPoint geoPoint = getGeoPoint(docId);
+      double distance =
+          GeoUtils.arcDistance(
+              origin.getLatitude(), origin.getLongitude(), geoPoint.getLat(), geoPoint.getLon());
+      double score = decayFunc.computeScore(distance, scale);
+      return Math.max(0.0, score - offset);
```

Comment:
I'm not sure this math is right. These are the function definitions https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html#_supported_decay_functions